From a2be82f0250ebe5ded00743c0435702104c17202 Mon Sep 17 00:00:00 2001
From: Michal Simek <michal.simek@xilinx.com>
Date: Mon, 26 Nov 2012 11:10:24 +0100
Subject: [PATCH 490/628] usb: xilinx: Enable ULPI phy detection for usb

git://github.com/Xilinx/linux-xlnx.git xilinx-v14.7
commit 7e3541236258a8048aa6f5c4e52cf93dab112f9e

Detect usb ulpi phys and set it up in pdata for others usb drivers.

Signed-off-by: Michal Simek <michal.simek@xilinx.com>
Signed-off-by: Liming Wang <liming.wang@windriver.com>
---
 drivers/usb/host/xusbps-dr-of.c |   20 +++++++++++++++++++-
 1 files changed, 19 insertions(+), 1 deletions(-)

diff --git a/drivers/usb/host/xusbps-dr-of.c b/drivers/usb/host/xusbps-dr-of.c
index bf26360..bc0e4af 100644
--- a/drivers/usb/host/xusbps-dr-of.c
+++ b/drivers/usb/host/xusbps-dr-of.c
@@ -27,6 +27,9 @@
 #include <linux/of_platform.h>
 #include <linux/string.h>
 #include <linux/clk.h>
+#include <linux/usb/ulpi.h>
+
+#include "ehci-xilinx-usbps.h"
 
 static u64 dma_mask = 0xFFFFFFF0;
 
@@ -145,7 +148,7 @@ static int __devinit xusbps_dr_of_probe(struct platform_device *ofdev)
 	const unsigned char *prop;
 	static unsigned int idx;
 	struct resource *res;
-	int i;
+	int i, phy_init;
 
 	pdata = &data;
 	memset(pdata, 0, sizeof(data));
@@ -184,6 +187,21 @@ static int __devinit xusbps_dr_of_probe(struct platform_device *ofdev)
 	prop = of_get_property(np, "phy_type", NULL);
 	pdata->phy_mode = determine_usb_phy(prop);
 
+	/* If ULPI phy type, set it up */
+	if (pdata->phy_mode == XUSBPS_USB2_PHY_ULPI) {
+		pdata->ulpi = otg_ulpi_create(&ulpi_viewport_access_ops,
+			ULPI_OTG_DRVVBUS | ULPI_OTG_DRVVBUS_EXT);
+		if (pdata->ulpi)
+			pdata->ulpi->io_priv = pdata->regs +
+							XUSBPS_SOC_USB_ULPIVP;
+
+		phy_init = usb_phy_init(pdata->ulpi);
+		if (phy_init) {
+			pr_info("Unable to init transceiver, missing?\n");
+			return -ENODEV;
+		}
+	}
+
 	for (i = 0; i < ARRAY_SIZE(dev_data->drivers); i++) {
 		if (!dev_data->drivers[i])
 			continue;
-- 
1.7.5.4

